### PR TITLE
Create the structure to handle ADRs

### PR DESCRIPTION
**Pull request by cescoffier**
- Create the structure as discussed on https://groups.google.com/d/msgid/quarkus-dev/CB395F5B-5A01-464C-8008-A0E8E7E247E6%40redhat.com?utm_medium=email&utm_source=footer
- Update the template to MADR

### CODE REVIEW

**Comment by gsmet** [comment]
I was wondering if it should be in a specific repo? Frankly the GH search engine is a massive pain and I'm afraid it will be hard to find a decision in the middle of the rest.

File: adr/README.adoc [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+== ADR lifecycle
+
+An ADR can have the following states:
+
+- _proposed_ - initial state
+- _accepted_ / _rejected_
+- _deprecated_ - The ADR is not applied anymore or is out of date
+- _superseded_ - another ADR replaces the current one. The ADR is updated with a link to the replacement.
+
+**IMPORTANT:** Once _accepted_, and ADR is immutable. Only its status and date can be updated.
```
Comment (gastaldi) [inline adr/README.adoc:26]:
```suggestion
**IMPORTANT:** Once _accepted_, an ADR is immutable. Only its status and date can be updated.
```

**Comment by gastaldi** [comment]
If these files are supposed to be here, better fix the CI to not run the full build too

**Comment by cescoffier** [comment]
I'm not against moving them to another repo. However, it would mean that only some people aware of such a repo will follow the PR and evolution. 

About the search engine, the file type may help, but I agree, not the best search engine.

File: adr/README.adoc (hunk 1 above)
Comment (cescoffier) [reply adr/README.adoc]: Fixed, thanks!

**Comment by cescoffier** [comment]
I've excluded the ADR directory from the build.

**Comment by FroMage** [comment]
To me they belong in this repo, and can be found by checking it out and using an IDE or file exporer, or by using the explorer view of github and Ctrl-F, no need to use the search box. Or are we expecting hundreds of such ADRs?

**Review by FroMage — CHANGES_REQUESTED** [review]
This mostly looks good, and I'm in favour of ADRs.

But I see no explanation as to what the process is that leads from `proposed` to either `accepted` or `rejected`.

Is it a vote? What sort? By whom? Time-boxed?

Same for `deprecated`: what's the process to deprecate an ADR? A PR changing its status? Where do we document why it was deprecated and what to use/do instead?

For superceded, I guess the process is that there's a new ADR in the same PR that supercedes the other, right? In the case of RFCs it helps tremendously if the new ADR links to the deprecated ones. We'll probably also have to document why it became superceded (in the old or new ADR?)

**Comment by cescoffier** [comment]
> But I see no explanation as to what the process is that leads from proposed to either accepted or rejected.
> Is it a vote? What sort? By whom? Time-boxed?

I agree we need some kind of process, but not a heavy one, that would kill everything. 
It should work like the other PRs, the relevant extension maintainer should discuss and decide the outcome. 

> Same for deprecated: what's the process to deprecate an ADR? A PR changing its status? Where do we document why it was deprecated and what to use/do instead?

That's a good question. It's actually interesting as no other ADR describe this process. The status is changed, but no reason given. We should include a reason.

> For superceded, I guess the process is that there's a new ADR in the same PR that supercedes the other, right? In the case of RFCs it helps tremendously if the new ADR links to the deprecated ones. We'll probably also have to document why it became superceded (in the old or new ADR?)

Yes, that should be the same PR, and the status links to the new document. 

I will update the PR with these changes.

**Comment by cescoffier** [comment]
@FroMage PR udpated. I've added more details about the process.

**Comment by gsmet** [comment]
I'm more talking about having pull requests lost in the middle of the rest. Not about being able to search files in a specific directory.

Please at least add a triage rule to attach a label to the PRs.

**Comment by cescoffier** [comment]
+1 for the label. I already created an area/adr label, so I just need to update the bot rules, right?

**Comment by cescoffier** [comment]
(BTW, I don't expect hundreds of ADR PRs, at least that's my hope)

**Comment by maxandersen** [comment]
+1 on this. just one suggestion:

I love *adoc* but may I suggest we use markdown for the .adrs as then we can use hackmd.io for collaborative editing (think google doc editing but without the google doc part) on these ? 

btw. what is MADR?

**Comment by cescoffier** [comment]
@maxandersen I'm not against markdown, but the lack of admonition and xref is a bit annoying. But I don't have a strong opinion (Except that I need to check the link syntax every single time).

MADR is the name of the structure I used as base: https://github.com/adr/madr

File: adr/README.adoc [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+- Create an asciidoc (`.adoc`) document in the _adr_ directory named "n+1-my-adr-title.adoc", with n+1 the id. Use 4 digits as id (0001, 0002...)
+- The document can be created by duplicating the _0000-template.adoc_ file. It is based on the https://adr.github.io/madr/[MADR].
+- The initial state of any ADR is _proposed_.
+- Once written, open a pull request to drive the discussion. Add the `adr` label to the pull request, as well as any related area labels.
+- Before merging the ADR pull request, update the state to _accepted_.
+- Email mailto:https://groups.google.com/g/quarkus-dev[Quarkus Dev Mailing List] to inform everyone about the new ADR.
+
+The discussion around an ADR happens on the pull request.
+When the participant agrees, the ADR is marked as _accepted_, and the PR is merged.
```
Comment (FroMage) [inline adr/README.adoc:18]:
```suggestion
When the participants agree, the ADR is marked as _accepted_, and the PR is merged.
```

**Review by FroMage — APPROVED** [review]

**Comment by FroMage** [comment]
LGTM, but I think more than just us four should vote on this.

**Review by maxandersen — CHANGES_REQUESTED** [review]
Missing the auto label rules as @gsmet suggested. 

With respect to .md we can just try hackmd on some as experiment and see if great or not worth jt. Let's not hold back this pr because of that.

**Comment by cescoffier** [comment]
I added the auto-label rule.

File: adr/README.adoc [hunk 3]
```diff
@@ -0,0 +1,40 @@
+= ADR
+
+https://adr.github.io/[ADR] stands _Architectural Decision Records_  and is a way to capture architectural decisions over time.
```
Comment (FroMage) [inline adr/README.adoc:3]:
```suggestion
https://adr.github.io/[ADR] stands for _Architectural Decision Records_  and is a way to capture architectural decisions over time.
```

**Review by FroMage — APPROVED** [review]

**Comment by cescoffier** [comment]
Moving forward, structure and process will be revisited if needed.

**Pull request merged**